feat: Implement CEL runtime cost budget for expression evaluation

Rudra-Sankha-Sinhamahapatra opened the pull request:
# PR Description

Fixes #191
## Feature Description
This PR implements cost budgeting for CEL expression evaluation to prevent resource exhaustion from complex expressions. The controller now enforces both per-call limits for individual expressions and a total runtime budget for the entire reconciliation cycle.

Key features implemented:
- Per-call cost limit (1,000,000 units, ~0.1 second of execution time)
- Total runtime budget (1,000 units for all expressions in a reconciliation cycle)
- Budget exceeded error handling and proper status reporting

## Implementation Details
- Added cost tracking constants and helper functions in `pkg/cel/cost.go`
- Implemented cost accumulation in `ResourceGraphDefinitionRuntime`
- Added error handling for budget exceeded cases in controller
- Updated expression evaluation to use cost tracking options

## Testing
- Added tests for CEL cost budget enforcement
- Verified error handling when budget is exceeded
- Added test cases with both simple and complex expressions

### `pkg/cel` Tests
<img width="802" alt="Screenshot 2025-03-22 at 5 44 34 PM" src="https://github.com/user-attachments/assets/4d493113-d83c-4bf2-a7ae-64a5909b522e" />

### `pkg/graph` Tests
<img width="802" alt="Screenshot 2025-03-22 at 5 44 51 PM" src="https://github.com/user-attachments/assets/98f53694-8530-44fa-bf3e-dd19511d1218" />

### `pkg/runtime` Tests
<img width="802" alt="Screenshot 2025-03-22 at 5 45 02 PM" src="https://github.com/user-attachments/assets/ef96dfc2-521d-48f3-b206-933190a2b273" />

### All tests
<img width="802" alt="Screenshot 2025-03-22 at 5 49 23 PM" src="https://github.com/user-attachments/assets/3dc30ad7-cb1f-4e8b-875d-ae4688d7856a" />

## Example Error
When the CEL budget is exceeded, the controller will report an error and the instance status will reflect:

```yaml
apiVersion: kro.run/v1alpha1
kind: ResourceGroup
metadata:
  name: my-application
status:
  conditions:
    - type: Ready
      status: "False"
      reason: CELBudgetExceeded
      message: "Total CEL cost exceeded budget of 1000"
```

This prevents potential DoS attacks from overly complex expressions and ensures the controller remains responsive.

cc: @a-hilaly

a-hilaly commented:
Thanks you @Rudra-Sankha-Sinhamahapatra - i'm curious to know what tools did you use to trigger the CEL budget exceeded error?

Rudra-Sankha-Sinhamahapatra commented:
I used Go's testing framework with -v flag to test it

Rudra-Sankha-Sinhamahapatra commented:
@a-hilaly I primarily used the go's testing framework with -v flag to trigger cek budget exceeded error 

in builder_test.go & cost_test.go  I have written tests to trigger it 

cost_test.go tests with basic cost tracking ,tests with zero budget, tests complex nested operations that exceeds budget

In builder_test.go it tests with three levels of expression from a simple expression,more complex expression & a highly complex expression  that will exceed cel budget cost 

<img width="690" alt="Screenshot 2025-04-06 at 2 40 06 PM" src="https://github.com/user-attachments/assets/a19a5438-28a7-4dde-9a4d-be9c06eadfdd" />

In runtime.go, I added cumulative cost tracking to monitor total CEL evaluation costs during resource processing

In controller_reconcile.go, I added error handling to show users when their expressions exceed the budget, which appears in their ResourceGroup status

jakobmoellerdev commented:
@Rudra-Sankha-Sinhamahapatra are you still interested in this? It would be amazing if you could rebase this feature and make it an "opt-in" in which people can opt-in to do cost evaluation

a-hilaly commented:
closing this PR due to innactivity please feel free to re-open it if you're still interrested in upstreaming this.

/close
